### ISSUE TITLE
Possible Integration of CZI via libCZI

### ISSUE THREAD

**Issue by zeissmicroscopy** [openslide/openslide#214]
Hi guys,

this is more a request not a real bug entry or issue.

Zeiss recently released libCZI:

https://github.com/zeiss-microscopy/libCZI

I know that OpenSlide did not want to use ZeissImgLib a while ago, but now we made this new library available on GitHub. It might be time to reconsider this decision?

Let me know, if you want to discuss this.

Sebi (from Zeiss)

**Comment by bgilbert** [comment]
Thank you for contributing this library to the open source community, and for bringing it to the project's attention.  Does Zeiss intend to actively maintain this library in future?

It's not immediately clear to me whether this is a direction OpenSlide should go.  Initial thoughts:

Pros:
- Zeiss will presumably do a better job supporting CZI than we could do ourselves.
- OpenSlide has generally tried to avoid reimplementing anything that's available in an existing library, in a suitable language, with a compatible license.

Cons:
- libCZI is GPLv3.  OpenSlide is LGPLv2.1, and there are existing users who depend on that license.  At a minimum we would need to make libCZI an optional dependency, since otherwise OpenSlide could not be used in GPLv2-only applications (if any exist in this space) or proprietary applications distributed to users.  (GPLv2+ code, and proprietary code running as a server backend, would be fine.)

  I notice that libCZI's README offers commercial licenses, presumably for those who don't want to abide by the GPLv3.  That's fine, but I'm not personally eager to be an instrument of that business model.
- OpenSlide is implemented in C, so there would need to be a C++ glue module that calls into libCZI.  I'm not opposed to that, and we've discussed it before in the context of DICOM, but it would require additional effort.

Possible approaches:
- Optionally support libCZI.  Someone would need to package libCZI for Fedora and Debian.  Linux distros would presumably include only one OpenSlide build, with libCZI included; this is probably fine for existing free software that uses OpenSlide or VIPS (I haven't exhaustively checked), and proprietary software probably doesn't link against distro libraries.  The Windows build would need to include two OpenSlide DLLs, with and without CZI support.
- Same, but dlopen libCZI if available.  That would avoid having to produce two Windows binaries, if it passes licensing muster; maybe Zeiss can comment on this.
- Implement our own driver (#144) instead of using libCZI.
- Optionally link against libCZI, falling back to our own driver if libCZI is not included.  This seems like the worst of both worlds: the built-in driver wouldn't get much use, and would exist only to support commercial users who would probably end up licensing libCZI anyway.
- Support neither approach, and point users to libCZI if they want CZI support.

I would welcome input from the community or from Zeiss.

**Comment by bgilbert** [comment]
@zeissmicroscopy As mentioned above, it would be helpful to know whether Zeiss intends to maintain libCZI with bug fixes, support for new CZI features, and so on.  In addition, does Zeiss intend to accept external contributions to libCZI?

**Comment by sebi06** [comment]
Yes, we have the clear commitment to maintain libCZI and fix bugs. The same is true if something fundamental changes inside the CZI data format itself.
We have not really talked about what to do in case of external contributions, be we are more than open to discuss as soon as there are any.
The main open question is still how to enable OpenSlide to read CZI. There are may requests for that.

**Comment by DerekJuba-NIST** [comment]
This is exactly what I am looking for in my current project. At this time, we are forced to convert the CZI file to a deepzoom format. If we can have direct integration of CZI into openslide, then we can create a tile processing server to host the tiles for visualization rather than having to deal with the additional disk write accesses. 

We attempted to use this repository: https://github.com/nsouedet/openslide, but received numerous errors (floating point exceptions and unable to convert tile data to rgba). Our current code is using libCZI for image processing, and is working well.

**Comment by jpfeil** [comment]
+1

**Comment by axakak** [comment]
+1

**Comment by mrcolo** [comment]
Same.
